x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/ScalarFunction.java

```diff
@@ -110,31 +116,56 @@ protected ScriptTemplate scriptWithScalar(ScalarFunction scalar) {
-                paramsBuilder().agg(aggregate).build(),
-                dataType());
+        Tuple<String, Integer> template = basicTemplate(aggregate);
+        return new ScriptTemplate(processScript(template.v1()),
+            repeatedParamBuilder(template.v2(), p -> p.agg(aggregate)).build(),
+            dataType());
     }
 
     // This method isn't actually used at the moment, since there is no grouping function (ie HISTOGRAM)
     // that currently results in a script being generated
     protected ScriptTemplate scriptWithGrouping(GroupingFunction grouping) {
-        String template = basicTemplate(grouping);
-        return new ScriptTemplate(processScript(template),
-                paramsBuilder().grouping(grouping).build(),
-                dataType());
+        Tuple<String, Integer> template = basicTemplate(grouping);
+        return new ScriptTemplate(processScript(template.v1()),
+            repeatedParamBuilder(template.v2(), p -> p.grouping(grouping)).build(),
+            dataType());
     }
 
     // FIXME: this needs to be refactored to account for different datatypes in different projects (ie DATE from SQL)
-    private String basicTemplate(Function function) {
+    private Tuple<String, Integer> basicTemplate(Function function) {
         if (function.dataType().name().equals("DATE") || function.dataType() == DATETIME ||
             // Aggregations on date_nanos are returned as string
             (function instanceof AggregateFunction && ((AggregateFunction) function).field().dataType() == DATETIME)) {
 
-            return "{sql}.asDateTime({})";
-        } else {
-            return "{}";
+            return new Tuple<>("{sql}.asDateTime({})", 1);
+        } else if (function instanceof AggregateFunction) {
+            DataType dt = function.dataType();
+            if (dt.isInteger()) {
+                // MAX, MIN, SUM need to retain field's data type, so that possible operations on integral types (like division) work
+                // correctly -> perform a cast in the aggs filtering script, the bucket selector for HAVING.
+                // SQL function classes not available in QL: filter by name
+                String fn = function.functionName();
+                if ("MAX".equals(fn) || "MIN".equals(fn)) {
+                    return new Tuple<>(safeCastNumericParamTemplate(dt), 2);
+                } else if ("SUM".equals(fn)) {
+                    return new Tuple<>(safeCastNumericParamTemplate(LONG), 2);
+                }
+            }
+        }
+        return new Tuple<>("{}", 1);
+    }
+
+    private static String safeCastNumericParamTemplate(DataType dataType) {
+        String type = dataType == INTEGER ? "int" : dataType.esType();
+        return "Double.NaN.compareTo({}) == 0 ? null : ((Number) {})." + type + "Value()";
+    }
+
+    private static ParamsBuilder repeatedParamBuilder(int count, Consumer<ParamsBuilder> consumer) {
```

Comment:
Why this approach with duplicating the _same_ parameter?
From what I tested, the query generated with the help of this method looks more or less like this (a snippet):

```
            "aggregations": {
                "4b4abc8": {
                    "stats": {
                        "field": "nr"
                    }
                },
                "having.beb296e4": {
                    "bucket_selector": {
                        "buckets_path": {
                            "a0": "4b4abc8.sum",
                            "a1": "4b4abc8.sum"
                        },
                        "script": {
                            "source": "InternalQlScriptUtils.nullSafeFilter(InternalQlScriptUtils.gt(InternalSqlScriptUtils.div(Double.NaN.compareTo(params.a0) == 0 ? null : ((Number) params.a1).longValue(),params.v0),params.v1))",
                            "lang": "painless",
                            "params": {
                                "v0": 2,
                                "v1": 5
                            }
                        },
                        "gap_policy": "skip"
                    }
                }
            }
```

Can we improve this?

```
                        "buckets_path": {
                            "a0": "4b4abc8.sum",
                            "a1": "4b4abc8.sum"
                        }
```
and use only one parameter both for the script source itself and for the value of that parameter? Something like the following:

```
            "aggregations": {
                "4b4abc8": {
                    "stats": {
                        "field": "nr"
                    }
                },
                "having.beb296e4": {
                    "bucket_selector": {
                        "buckets_path": {
                            "a0": "4b4abc8.sum"
                        },
                        "script": {
                            "source": "InternalQlScriptUtils.nullSafeFilter(InternalQlScriptUtils.gt(InternalSqlScriptUtils.div(Double.NaN.compareTo(params.a0) == 0 ? null : ((Number) params.a0).longValue(),params.v0),params.v1))",
                            "lang": "painless",
                            "params": {
                                "v0": 2,
                                "v1": 5
                            }
                        },
                        "gap_policy": "skip"
                    }
                }
            }
```

Reply:
> Can we improve this?

Not too elegantly, since the script binding is done with `format()`, which requires an actual extra parameter (which then generates another bucket path).
But I've replaced this implementation with an internal script function that just wraps the parameter (so only one parameter is provided, just like originally done).